Implement FR - Issue#301

Thinkersbluff opened the pull request:
Start the Bed Mesh Validation Pattern with one of the Full Circle pads instead of the front left corner.

### Description
Starts the Bed Mesh Validation Pattern with one of the Full Circle pads instead of the front left corner.

Achieves this by modifying line#72 in MeshValidationHandler.cpp, to modify the start position parameters of the G26 command.

Since G26 will automatically locate and move to the actual probe position closest to the specified start point to start the pattern, this same change should also work for the CR6-MAX.


### Requirements

No special requirements. 
This change should be fully compatible with all configurations of CR6 printer that is using the CR6Community display firmware.

### Benefits

See Issue#301.

### Configurations

Use Release Final 6.1 configs.

### Related Issues

(https://github.com/CR6Community/Marlin/issues/301)

Thinkersbluff committed:
Implement FR - Issue#301

Start the Bed Mesh Validation Pattern with one of the Full Circle pads instead of the front left corner.

Sebazzz commented:
Can you use the bed size divided by 2? Then you should be in the center, also for the CR-6 MAX.

Thinkersbluff commented:
I don't know how to code.  

I thought it would be good enough to get G26 to start on any circular pad.  Using 117,117 is my arbitrary way of nudging G26 closer to one of the circles.
From a "coding" point of view, it was within my limited skill set and it seemed to present the least risk of me introducing a programming error.

The middle of the MAX bed is at 200,200. 
As long as the MAX ABL mesh is at least 4x4, then G26 will start with the same circular pad on the MAX as it does on the SE, if sent to X117,Y117 to start, because 117 is closer to 133 than it is to 0.

If you prefer to implement this with includes, assigns, etc, I will need some more verbose guidance or I will need to ask that this issue please be reassigned to a code-capable resource.

Thinkersbluff commented:
OK.  
Now that the initial panic is over and my coffee is kicking-in, let's try me learning how to code enough to change this to using X_BED_SIZE/2, Y_BED_SIZE/2, instead of 117, 117.

Looking at AxisSettingsHandler.cpp, I think I see how to do this, but I am like a parrot, now, I might sound like I am using words, but I have no idea what they mean (yet?)  I have an appointment this morning, so do not have time to compile and test the change.  Please give me this day to figure it out. I will update this thread when I have a modified version to discuss.

RWBRITT commented:
Wouldn't bed size for x and y be defined in configuration file? Could you not take those variables and reference them in the file you are modifying and divide by 2?

Thinkersbluff commented:
Like this?

sprintf_P(gcodeBuffer, PSTR("G90\nG0 X0\nG26 B%d H%d P15 R Q1 X%d Y%d"), bed_temperature, nozzle_temperature, X_BED_SIZE / 2, Y_BED_SIZE / 2);

Thinkersbluff committed:
Use bed_size/2 instead of 117

Makes the code more readable and ensures same operation of bed validation pattern function on any bed size. More elegant and future-proof.

Thinkersbluff commented:
I just compiled and tested the revised change, and it works.

Thinkersbluff commented:
![image](https://user-images.githubusercontent.com/36551518/151709569-873e5148-4646-4518-b02b-a3b50a177491.jpeg)

![image](https://user-images.githubusercontent.com/36551518/151709626-3c745736-e6a9-47b7-b0d9-6e22553f29bf.jpeg)

Sebazzz commented:
Awesome! You've already merged something in, is that correct?

Thinkersbluff commented:
I have not commited it to the master extui branch, just compiled and tested my own copy.

Thinkersbluff commented:
BTW, as I posted on the Discord (https://discord.com/channels/759603746270609428/829457315454255204/937464385801179156), on my system, the tiny round imprints of the nozzle's nose only appear at the end of each arc when the Z Offset is at what I assume is exactly the best value.  Even 0.01mm up or down results in "tearing" of the filament at the end of each stroke.
These pics show the back of the last pad drawn, at each of three offsets:
![image](https://user-images.githubusercontent.com/36551518/151813985-20b5660c-b425-45c6-8c73-2422a42a41a2.jpeg)
![image](https://user-images.githubusercontent.com/36551518/151814472-27d65fe6-e238-4121-a46a-21f861a7d1a5.jpeg)
![image](https://user-images.githubusercontent.com/36551518/151814541-57c2b696-98bc-46cf-ad39-b98355b717d4.jpeg)
If that also occurs for others, that would seem to make the changes I propose very interesting for our members who wonder whether their rainbow-coloured bed mesh display needs corrective action on their part. That is, after-all, the purpose of G26, according to Marlin's documentation.

Thinkersbluff commented:
Of course, the 3rd change I made, which is not yet proposed, pending the outcome of #301 and #302, is to change Q4 to Q%d, and to use the Q retraction factor defined in Configuration.h instead of 4.
That is what presently allows me to draw the pattern with my Dragon hotend.

I am ready to submit that pull request too, but waiting to hear whether these changes are acceptable, before piling-on another one.

Thinkersbluff commented:
Question, @Sebazzz  - The automated test scripts are failing to validate this PR.  

I am seeing, "Write-FatalError: /home/runner/work/Marlin/Marlin/scripts/Run-ExampleConfigBuilds.ps1:130
Line |
 130 |          Write-FatalError "Unable to find platform.io environment name …
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unable to find platform.io environment name for anycubic-vyper"

Do I need to add something to the PR to accomodate anycubic-vyper?

Sebazzz commented:
LGTM!

@Thinkersbluff Not sure why the pio build fails - I'Il take a look at it!
